hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/common/ChunkBuffer.java

```diff
@@ -44,9 +45,6 @@ static ChunkBuffer allocate(int capacity) {
    *   When increment <= 0, entire buffer is allocated in the beginning.
    */
   static ChunkBuffer allocate(int capacity, int increment) {
-    if (increment > 0 && increment < capacity) {
-      return new IncrementalChunkBuffer(capacity, increment, false);
```

Comment:
Can you please also change `TestChunkBuffer#runTestIncrementalChunkBuffer` to explicitly create `IncrementalChunkBuffer`?  Currently it uses this factory method, and so with this patch it really tests `ChunkBufferImplWithByteBuffer`.

Reply:
Wow, nice catch.

Reply:
IncrementalChunkBuffer was added to address cases where ozone client were running into OOM with keys less than chunk size , as without this, the smallest buffer which will be allocated will always be equal to the chunk size(4MB by default).

Please see https://issues.apache.org/jira/browse/HDDS-2331 for more details.
I would prefer to not remove this logic of incremental chunk buffer and may be hide it within an internal config.

@elek , how much of perf gain we will have of we still do incremental buffer allocation?

Reply:
> @elek , how much of perf gain we will have of we still do incremental buffer allocation?

I can repeat the test to get exact numbers, but I couldn't get good performance without removing the incremental buffer. You can easily test it with the new unit test, if you write a lot of data with byte=1, it's still low.

> IncrementalChunkBuffer was added to address cases where ozone client were running into OOM with keys less than chunk size , as without this, the smallest buffer which will be allocated will always be equal to the chunk size(4MB by default).

I think it's a valid (and important question), but as far as I see it's safe to remove the `IncrementalByteBuffer`. As far as I see the situation is slightly different since HDDS-2331. I tried to test this patch with the commands from the HDDS-2331:

```
ozone freon rk --numOfThreads 1 --numOfVolumes 1 --numOfBuckets 1 --replicationType RATIS --factor ONE --keySize 1048576 --numOfKeys 5120 --bufferSize 65536
```

I couldn't reproduce the OOM.

Based on my understanding:

 1. We already have an increment by the `ByteBuffer` but size of the increment is 4MB (adding one more buffer when required)
 2. 4MB seems to be acceptable even with many clients in the same JVM, especially if we can have acceptable performance.

Let's say I have 100 Ozone clients (in the same JVM!!!) which write 1kb keys. I will have (4MB-1kb)  *100 overhead without the `IncrementalChunkBuffer` (as far as I understood). It's still <400MB in exchange for 30-100% performance gain. Sounds like a good deal.

But let me know if you see any problems here.

3. Let's say the 400MB overhead is unacceptable (or my calculation was wrong and the overhead is higher ;-) )

As far as I see the `BufferPool` is created per key. I think it would be possible to set the buffer size to `min(keySize, bufferSize)`. With this approach the first and only buffer of the BufferPool can have exactly the required size (which covers all the where the key size is < 4MB)

Reply:
> the situation is slightly different since HDDS-2331

Note that default chunk size was 16MB at the time when HDDS-2331 was reported.  The benefit from IncrementalChunkBuffer is less now with 4MB default size.

Reply:
I would suggest separating the two parts of the PR:

 1. reorganize the position calculation and allocation
 2. remove the usage of the Incremental buffer

While we can continue to searching for the safest method to do 2 (or do something instead of the removal), we can merge the first part where we already have an agreement.

Reply:
See #1374 about the 2nd.